### PR TITLE
Performance improvement for isFileBlacklisted by using caching

### PR DESCRIPTION
**Pull request by mmattel**
This PR improves the performance of `isFileBlacklisted` by using caching.
`isFileBlacklisted` is called from many places and used quite often, especially from scanner.php

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
> This PR improves the performance of isFileBlacklisted by using caching.

Can we please get a proof for this - e.g. send a comparison graph from blackfire.io

THX

**Comment by LukasReschke** [comment]
# Without memcache (100'000 runs)
With this change: 3.266398191452 sec
Without this change: 1.0600299835205 sec

=> Getting 2.2 secs *slower*

# With APCu enabled (100'000 runs)
With this change: 2.5553448200226 sec
Without this change: 1.0700287818909 sec

=>  Getting 1.5 secs *slower*
<hr/>
It actually makes things quite slower because querying the container is a quite expensive operation.

Please notice that actual execution time in prod. environments is faster than this measures since I had XDebug enabled.

**Comment by LukasReschke** [comment]
Test script:

```php
	for ($x = 0; $x <= 100000; $x++) {
		\OC\Files\Filesystem::isFileBlacklisted('aLongRandomFileName123');
	}

	$time2 = microtime(true);
	echo "script execution time: ".($time2-$time1); //value in seconds
```

File: lib/private/files/filesystem.php [hunk 1]
```diff
@@ -567,11 +567,16 @@ static public function isBlacklisted($data) {
...
-		$blacklist = \OC_Config::getValue('blacklisted_files', array('.htaccess'));
-		$filename = strtolower(basename($filename));
-		return in_array($filename, $blacklist);
+		$filename = strtolower(basename(self::normalizePath($filename)));
+		$memcache = \OC::$server->getMemCacheFactory()->create('isFileBlacklisted');
+		if ($memcache->hasKey('blacklisted_file')) {
+			$blacklist = $memcache->get('blacklisted_file');
+		}
+		else {
+			$blacklist = array_map('trim', \OC::$server->getSystemConfig()->getValue('blacklisted_files', array('.htaccess')));
```
Comment (nickvergessen) [inline lib/private/files/filesystem.php]: We can increase the performance, if we store `\OC::$server->getSystemConfig()->getValue('blacklisted_files', array('.htaccess'))` as a class member.
This is also what I suggested in the other PR

Comment (nickvergessen) [inline lib/private/files/filesystem.php]: Also the trim() in the array_map here is a behaviour change

Comment (DeepDiver1975) [inline lib/private/files/filesystem.php]: > We can increase the performance, if we store \OC::$server->getSystemConfig()->getValue('blacklisted_files', array('.htaccess')) as a class member.

the config class already caches values in RAM - I see no need in further fine tuning at this end.

**Pull request closed without merging**